open_spiel/games/phantom_go/phantom_go_board.cc

```diff
@@ -0,0 +1,868 @@
+  board_[largest_chain_head].chain_next = p;
+  board_[p].chain_head = largest_chain_head;
+  chain(largest_chain_head).num_stones += 1;
+
+  Neighbours(p, [this, largest_chain_head](VirtualPoint n) {
+    if (IsEmpty(n)) {
+      chain(largest_chain_head).add_liberty(n);
+    }
+  });
+}
+
+void GoBoard::RemoveLibertyFromNeighbouringChains(VirtualPoint p) {
+  Neighbours(p, [this, p](VirtualPoint n) { chain(n).remove_liberty(p); });
+}
+
+int GoBoard::CaptureDeadChains(VirtualPoint p, GoColor c) {
+  int stones_captured = 0;
+  int capture_index = 0;
+  Neighbours(p, [this, c, &capture_index, &stones_captured](VirtualPoint n) {
+    if (PointColor(n) == OppColor(c) && chain(n).num_pseudo_liberties == 0) {
+      last_captures_[capture_index++] = ChainHead(n);
+      stones_captured += chain(n).num_stones;
+      RemoveChain(n);
+    }
+  });
+
+  for (; capture_index < last_captures_.size(); ++capture_index) {
+    last_captures_[capture_index] = kInvalidPoint;
+  }
+
+  return stones_captured;
+}
+
+void GoBoard::RemoveChain(VirtualPoint p) {
+  VirtualPoint this_chain_head = ChainHead(p);
+  VirtualPoint cur = p;
+  do {
+    VirtualPoint next = board_[cur].chain_next;
+
+    SetStone(cur, GoColor::kEmpty);
+    InitNewChain(cur);
+
+    Neighbours(cur, [this, this_chain_head, cur](VirtualPoint n) {
+      if (ChainHead(n) != this_chain_head || IsEmpty(n)) {
+        chain(n).add_liberty(cur);
+      }
+    });
+
+    cur = next;
+  } while (cur != p);
+}
+
+void GoBoard::InitNewChain(VirtualPoint p) {
+  board_[p].chain_head = p;
+  board_[p].chain_next = p;
+
+  Chain& c = chain(p);
+  c.reset();
+  c.num_stones += 1;
+
+  Neighbours(p, [this, &c](VirtualPoint n) {
+    if (IsEmpty(n)) {
+      c.add_liberty(n);
+    }
+  });
+}
+
+bool GoBoard::IsInBoardArea(VirtualPoint p) const {
+  auto rc = VirtualPointTo2DPoint(p);
+  return rc.first >= 0 && rc.first < board_size() && rc.second >= 0 &&
+         rc.second < board_size();
+}
+
+bool GoBoard::IsLegalMoveObserver(VirtualPoint p, GoColor c) const {
+  if (p == kVirtualPass) return true;
+  if (!IsInBoardArea(p)) return false;
+  if (!IsEmpty(p) || p == LastKoPoint()) return false;
+  if (chain(p).num_pseudo_liberties > 0) return true;
+
+  // For all checks below, the newly placed stone is completely surrounded by
+  // enemy and friendly stones.
+
+  // Allow to play if the placed stones connects to a group that still has at
+  // least one other liberty after connecting.
+  bool has_liberty = false;
+  Neighbours(p, [this, c, &has_liberty](VirtualPoint n) {
+    has_liberty |= (PointColor(n) == c && !chain(n).in_atari());
+  });
+  if (has_liberty) return true;
+
+  // Allow to play if the placed stone will kill at least one group.
+  bool kills_group = false;
+  Neighbours(p, [this, c, &kills_group](VirtualPoint n) {
+    kills_group |= (PointColor(n) == OppColor(c) && chain(n).in_atari());
+  });
+  if (kills_group) return true;
+
+  return false;
+}
+
+//update 1
+//finish or rework
```

Comment:
Is this not finished?

Reply:
The method is finished, it is just residual comment which was left from beginnings of the project.